src/web/hooks/use-call-tool.ts

```diff
@@ -1,51 +1,86 @@
 import { useState } from "react";
 import type { CallToolArgs, CallToolResponse } from "../types.js";
 
-type BaseCallToolState<
-  TStatus extends "idle" | "pending" | "success" | "error",
-  TData extends CallToolResponse = CallToolResponse
-> = {
-  status: TStatus;
-  isIdle: TStatus extends "idle" ? true : false;
-  isPending: TStatus extends "pending" ? true : false;
-  isSuccess: TStatus extends "success" ? true : false;
-  isError: TStatus extends "error" ? true : false;
-  data: TStatus extends "success" ? TData : undefined;
-  error: TStatus extends "error" ? unknown : undefined;
+type CallToolState<TData extends CallToolResponse = CallToolResponse> =
+  | {
+      status: "idle";
+      isIdle: true;
+      isPending: false;
+      isSuccess: false;
+      isError: false;
+      data: undefined;
+      error: undefined;
+    }
+  | {
+      status: "pending";
+      isIdle: false;
+      isPending: true;
+      isSuccess: false;
+      isError: false;
+      data: undefined;
+      error: undefined;
+    }
+  | {
+      status: "success";
+      isIdle: false;
+      isPending: false;
+      isSuccess: true;
+      isError: false;
+      data: TData;
+      error: undefined;
+    }
+  | {
+      status: "error";
+      isIdle: false;
+      isPending: false;
+      isSuccess: false;
+      isError: true;
+      data: undefined;
+      error: unknown;
+    };
+
+type SideEffects<ToolArgs, ToolResponse> = {
+  onSuccess?: (data: ToolResponse, toolArgs: ToolArgs) => void;
+  onError?: (error: unknown, toolArgs: ToolArgs) => void;
+  onSettled?: (
+    data: ToolResponse | undefined,
+    error: unknown | undefined,
+    toolArgs: ToolArgs
+  ) => void;
 };
 
-type IdleCallToolState = BaseCallToolState<"idle">;
-type PendingCallToolState = BaseCallToolState<"pending">;
-type SuccessCallToolState<TData extends CallToolResponse = CallToolResponse> =
-  BaseCallToolState<"success", TData>;
-type ErrorCallToolState = BaseCallToolState<"error">;
+type CallToolAsyncFn<TA, TR> = TA extends null
+  ? () => Promise<TR>
+  : (toolArgs: TA) => Promise<TR>;
 
-type CallToolState<TData extends CallToolResponse = CallToolResponse> =
-  | IdleCallToolState
-  | PendingCallToolState
-  | SuccessCallToolState<TData>
-  | ErrorCallToolState;
+type CallToolResponseConstraint = Partial<
+  Pick<CallToolResponse, "structuredContent" | "meta">
+>;
 
 export const useCallTool = <
   ToolArgs extends CallToolArgs = null,
-  ToolResponse extends CallToolResponse = CallToolResponse
+  ToolResponse extends CallToolResponseConstraint = CallToolResponseConstraint
 >(
```

Comment:
For coherence with useToolInfo type :)?


```suggestion
type ToolResponseSignature = Pick<CallToolResponse, "structuredContent" | "meta">;

export const useCallTool = <
  ToolArgs extends CallToolArgs = null,
  ToolResponse extends Partial<ToolResponseSignature> = {}
>(
```